### PR TITLE
Scheduled workflows

### PR DESCRIPTION
**Pull request by chuck-dbos**

### CODE REVIEW

File: docs/tutorials/scheduled-workflows.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        Running a workflow every 5 seconds -
+          scheduled time: ${schedTime.toISOString()} /
+          actual start time: ${startTime.toISOString()}
+    `);
+  }
+}
+```
+
+### Scheduling The Workflow
+Then, annotate your method with a [`@Scheduled`](../api-reference/decorators.md#scheduled) decorator specifiying the schedule in a [`crontab`](https://en.wikipedia.org/wiki/Cron)-like format, which optionally specifies a constraint for seconds, followed by constraints for minutes, hours, day of month, month of year, and day of week.  For example, the following workflow will run evey 30 seconds:
```
Comment (kraftp) [inline docs/tutorials/scheduled-workflows.md:31]: Typo (evey)

File: docs/tutorials/scheduled-workflows.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+description: Learn how to run DBOS workflows on a schedule.
+---
+
+In this guide, you'll learn how to execute DBOS workflows exactly once per time interval.
+
+### Defining A Workflow Function For Scheduling
+First, define your workflow.  Its parameters must be:
+- The `WorkflowContext`
+- The time that the run was scheduled (as a `Date`).
+- The time that the run was actually started (as a `Date`).  This can be used to tell if an exacty-once workflow was started behind schedule.
```
Comment (kraftp) [inline docs/tutorials/scheduled-workflows.md:13]: Typo (exacty)

File: docs/api-reference/decorators.md [hunk 3]
```diff
@@ -544,6 +545,54 @@ class KafkaExample{
 }
 ```
 
+### Scheduled Workflow Decorators
+
+#### `@Scheduled(schedulerConfig: SchedulerConfig)` {#scheduled}
+
+Runs a workflow function on a specified schedule, with guarantees such as executing exactly once per scheduled interval.
+
+By default, the workflow is executed exactly once per scheduled interval.  This means executions will be started concurrently and can overlap, and that if the application is taken down and restarted, makeup work will be performed.  Tracking of execution history is implemented using system database records.  An workflow key (consisting of the workflow function name and scheduled time) is used to deduplicate any workflows that may inadvertently be initiated by the scheduler.
```
Comment (qianl15) [inline docs/api-reference/decorators.md:554]: Here we can directly say the idempotency key of a scheduled workflow is composed of the workflow function name and scheduled time.

```suggestion
By default, the workflow is executed exactly once per scheduled interval.  This means executions might be started concurrently and can overlap, and that if the application is taken down and restarted, makeup work will be performed.  Tracking of execution history is implemented using system database records.  An workflow idempotency key (consisting of the workflow function name and scheduled time) is used to deduplicate any workflows that may inadvertently be initiated by the scheduler.
```

File: docs/api-reference/decorators.md [hunk 4]
```diff
@@ -544,6 +545,54 @@ class KafkaExample{
...
+
+The schedule is specified in a format similar to a traditional [`crontab`](https://en.wikipedia.org/wiki/Cron), with the following notes:
```
Comment (qianl15) [inline docs/api-reference/decorators.md:556]: Are those additional rules the same as [node-cron](https://www.npmjs.com/package/node-cron)? If so, let's put a reference to node-cron here.

File: docs/api-reference/decorators.md [hunk 5]
```diff
@@ -544,6 +545,54 @@ class KafkaExample{
 }
 ```
 
+### Scheduled Workflow Decorators
+
+#### `@Scheduled(schedulerConfig: SchedulerConfig)` {#scheduled}
+
+Runs a workflow function on a specified schedule, with guarantees such as executing exactly once per scheduled interval.
+
+By default, the workflow is executed exactly once per scheduled interval while the application is active.  This means executions might be started concurrently and can overlap, and that if the application is taken down and restarted, makeup work will be performed.  Tracking of execution history is implemented using system database records.  A workflow idempotency key (consisting of the workflow function name and scheduled time) is used to deduplicate any workflows that may inadvertently be initiated by the scheduler.
```
Comment (qianl15) [inline docs/api-reference/decorators.md:554]:
```suggestion
By default, the workflow is executed exactly once per scheduled interval.  This means executions might be started concurrently and can overlap, and that if the application is taken down and restarted, makeup work will be performed.  Tracking of execution history is implemented using system database records.  A workflow idempotency key (consisting of the workflow function name and scheduled time) is used to deduplicate any workflows that may inadvertently be initiated by the scheduler.
```

File: docs/tutorials/scheduled-workflows.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+class ScheduledExample{
+  @Workflow()
+  @Scheduled({crontab: '*/30 * * * * *'})
+  static async scheduledFunc(ctxt: WorkflowContext, schedTime: Date, startTime: Date) {
+    ctxt.logger.info(``);
+  }
+}
+```
+
+The DBOS crontab format supports some common extensions, as seen in the following examples:
```
Comment (qianl15) [inline docs/tutorials/scheduled-workflows.md:45]: Put a pointer to the full specification?

File: docs/tutorials/scheduled-workflows.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
+---
+sidebar_position: 18
+title: Scheduling Workflows
```
Comment (qianl15) [inline docs/tutorials/scheduled-workflows.md:3]:
```suggestion
title: Scheduled Workflows
```

File: docs/tutorials/scheduled-workflows.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+- `30 * * * * *`: 30 seconds after the beginning of every minute
+- `0 12 * * Sun`: At noon on sundays
+- `0 9-17 * * *`: At the top of the hour, from 9am to 5pm
+- `*/30 * * * * *`: Every 30 seconds (seconds is evenly divisible by 30)
+- `10-19/2 * * January,Feb *`: At 10, 12, 14, 16, and 18 minutes into each hour of every day, only in the months of January and February
+
+### How Scheduling Works
+Under the hood, DBOS constructs an [idempotency key](./idempotency-tutorial) for each workflow invocation.  The key is a concatenation of the function name and the scheduled time, ensuring each scheduled invocation occurs at most once.  The DBOS system database tracks the last time that the workflow was known to have started; this implementation ensures that the workflow is started at least once.  The combination of mechanisms can therefore ensure that the scheduled workflow occurs exactly once per time interval.
+
+### Skipping scheduled workflows
```
Comment (qianl15) [inline docs/tutorials/scheduled-workflows.md:57]:
```suggestion
### Skipping Scheduled Workflows
```

File: docs/tutorials/scheduled-workflows.md (hunk 6 above)
Comment (chuck-dbos) [reply docs/tutorials/scheduled-workflows.md:45]: I haven't found one.  I have one that I'm pasting in shortly.

File: docs/tutorials/scheduled-workflows.md [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+### Skipping Scheduled Workflows
+By default, scheduled workflows are executed exactly once per interval.  This means that, if an application is stopped for an extended period, DBOS Transact could initiate a significant number of workflows upon restart.  If it is known by the application developer that "makeup work" should not be scheduled for a given workflow function, the mode for that function should be changed in its configuration:
+```typescript
+    @Scheduled({mode: SchedulerMode.ExactlyOncePerIntervalWhenActive, crontab: '...'})
+```
+
+If the mode is set to `ExactlyOncePerIntervalWhenActive` , DBOS Transact will not consult the system database to figure out if scheduled workflows were missed, but will instead start the schedule at the current time.
+
+### `crontab` Specification
```
Comment (kraftp) [inline docs/tutorials/scheduled-workflows.md:65]: Maybe make it clear that this is a tutorial on a well-known format and not something we invented, and link to other helpful resources

File: docs/api-reference/decorators.md [hunk 10]
```diff
@@ -525,6 +526,54 @@ class KafkaExample{
 }
 ```
 
+### Scheduled Workflow Decorators
+
+#### `@Scheduled(schedulerConfig: SchedulerConfig)` {#scheduled}
+
+Runs a workflow function on a specified schedule, with guarantees such as executing exactly once per scheduled interval.
+
+By default, the workflow is executed exactly once per scheduled interval.  This means executions might be started concurrently and can overlap, and that if the application is taken down and restarted, makeup work will be performed.  Tracking of execution history is implemented using system database records.  A workflow idempotency key (consisting of the workflow function name and scheduled time) is used to deduplicate any workflows that may inadvertently be initiated by the scheduler.
```
Comment (kraftp) [inline docs/api-reference/decorators.md:535]: This paragraph is a bit verbose.  Some detail can be omitted.

File: docs/tutorials/scheduled-workflows.md (hunk 9 above)
Comment (chuck-dbos) [reply docs/tutorials/scheduled-workflows.md:65]: I had tried to do that at the top.  Qian wanted a link to a full specification, which I never found.  I can move it from the tutorial to the reference section, however.  Maybe that's more appropriate?

**Pull request merged**